### PR TITLE
enhm/optimisations

### LINKED ISSUE #36: Limit Runtime Costs of Tracing
When encountering repeating structures, such as for loops, while loops, where the body of the loop is executed repeatedly, usually the types will not change. Tracing can be turned off for these structures, as they will just duplicate information.

### PR DESCRIPTION
**Pull request by bengsparks**
The tests currently fail in the implementation in an area where I will need help solving it.
But the tests for the optimisations pass, meaning the feature in itself is stable enough for merging after this is rectified.

Closes #36

### CODE REVIEW

File: tests/tracing/optimisation/test_loop.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    vars = [
+        ("s", int),
+        ("x", int),
+        ("r", int),
+        (skippable_looping.__name__, int),
+    ]
+    assert df.shape[0] == len(vars)
+
+    for (name, _) in vars:
```
Comment (RecurvedBow) [inline tests/tracing/optimisation/test_loop.py:39]: From my understanding, the goal of the test is to prove that the optimisation works. However, even without any loop optimization, the test would still pass. You need to add asserts that pass with optimizations, but fail without.

File: tracing/optimisation/base.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
+from abc import abstractmethod
+
+from .enums import TriggerStatus
+from .utils import FrameWithMetadata
+
+import pandas as pd
+
+
+class Optimisation:
```
Comment (RecurvedBow) [inline tracing/optimisation/base.py:9]: Add ABC as base class as there are abstract methods.

File: tracing/optimisation/enums.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import enum
+
+
+class TriggerStatus(enum.Enum):
```
Comment (RecurvedBow) [inline tracing/optimisation/enums.py:4]: Change to IntEnum.

File: tracing/optimisation/utils.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+from traceback import linecache # type: ignore
+import functools
+import io
+import typing
+import tokenize
+
+
+# Start with characters or underscore, rest is the same plus numbers
+@dataclass
+class FrameWithMetadata:
```
Comment (RecurvedBow) [inline tracing/optimisation/utils.py:11]: Why is there no constructor?

Comment (RecurvedBow) [inline tracing/optimisation/utils.py:11]: Isn't it more clean to define a function get_frame_metadata_from(frame) -> FrameMetaData, which contains the properties?

Comment (RecurvedBow) [inline tracing/optimisation/utils.py:11]: There aren't any tests to validate the functions. Create a test in which a function with loops is traced and use the frames to generate the meta data, and validate the function return values.

File: tracing/tracer.py [hunk 5]
```diff
@@ -19,21 +31,29 @@ def __init__(self, project_dir: pathlib.Path):
...
         self.project_dir = project_dir
-        self.old_values_by_variable_by_function_name: dict[str, dict] = {}
-        self._reset_members()
+
+        self.optimisation_stack: list[Optimisation] = list()
+        self.old_values_by_variable_by_function_name: dict[str, dict] = dict()
 
     def start_trace(self) -> None:
-        """Resets the trace values and starts the trace."""
-        self._reset_members()
```
Comment (RecurvedBow) [inline tracing/tracer.py:27]: Why has it been removed?

File: tracing/tracer.py [hunk 6]
```diff
@@ -85,65 +141,89 @@ def _evaluate_object(self, class_object: typing.Any) -> dict[str, type]:
...
+            logger.info(f"Skipping exception: {inspect.getframeinfo(frame)}")
             pass
 
-        else:
-            self.stop_trace()
-            print("The event " + str(event) + " is unknown.")
-            # Note: The value error does not stop the program for some reason.
-            raise ValueError("The event" + str(event) + " is unknown.")
+        # NOTE: If there is any error occurred in the trace function, it will be unset, just like settrace(None) is called.
+        # NOTE: therefore, throwing an exception does not work, as the trace function will simply be unset
```
Comment (RecurvedBow) [inline tracing/tracer.py:200]: While the comment is true, the code is still neccessary to determine whether the tracer has encountered an unhandled event (= default-case in switch). Add the else and remove the #Note in the else.

File: tracing/tracer.py [hunk 7]
```diff
@@ -43,12 +63,48 @@ def active_trace(self) -> typing.Iterator[None]:
...
+            logger.debug(
+                f"Applying TypeStableLoop for {inspect.getframeinfo(fwm._frame)}"
+            )
+            tsl = TypeStableLoop(fwm)
+            if not self.optimisation_stack or tsl != self.optimisation_stack[-1]:
+                self.optimisation_stack.append(tsl)
+                return
+
+    def _advance_optimisations(self, fwm: FrameWithMetadata) -> None:
+        for optimisation in self.optimisation_stack:
```
Comment (RecurvedBow) [inline tracing/tracer.py:106]: Why is it a stack if it is used as a queue?

File: tracing/optimisation/ignore.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    Dumb optimisation that is always ONGOING, in order to unconditionally
+    disable tracing for the relevant scope.
+    """
+
+    def __init__(self, frame: utils.FrameWithMetadata):
+        super().__init__(frame)
+        # By default, this optimisation is constantly active, as long as the current scope is active
+        self._status = TriggerStatus.ONGOING
+
+    def advance(self, current_frame: utils.FrameWithMetadata, _: pd.DataFrame) -> None:
```
Comment (RecurvedBow) [inline tracing/optimisation/ignore.py:20]: Change to update status. I would have assumed that data is passed from one advance to the next.

**Review by RecurvedBow — CHANGES_REQUESTED** [review]
Judging from the algorithm and its complexity, I would actually assume that the change does NOT improve the performance of the tracer. To counter-prove this, add a test with lots of iterations and show that with optimizations, the average time to run the function is significantly less than without.

File: tests/tracing/optimisation/test_ignored.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    p2 = pathlib.Path(
+        "another", "really", "cool", "file.backup"
+    )  # Tracing is turned off for this constructor now
+
+    d = p.is_dir()  # Do not trace
+
+    return p.is_file() - bool(b) + d  # Call is not traced, return type is
+
+
+def test_pathlib_calls_are_not_traced():
```
[the remark this answers is not in the document]
Comment (bengsparks) [reply tests/tracing/optimisation/test_ignored.py:24]: To see that all variables are still traced.

File: tracing/optimisation/utils.py (hunk 4 above)
Comment (bengsparks) [reply tracing/optimisation/utils.py:11]: The optimisations need both due to hidden mutable references. This cost me three days of debugging, which is why I put everything in this class

File: tracing/optimisation/base.py (hunk 2 above)
Comment (bengsparks) [reply tracing/optimisation/base.py:9]: Good point.

File: tracing/optimisation/enums.py (hunk 3 above)
Comment (bengsparks) [reply tracing/optimisation/enums.py:4]: TIL.

File: tracing/tracer.py (hunk 7 above)
Comment (bengsparks) [reply tracing/tracer.py:106]: It is used as a queue. Check `_update_optimisations`, which pops elements from the top.

File: tracing/optimisation/ignore.py (hunk 8 above)
Comment (bengsparks) [reply tracing/optimisation/ignore.py:20]: `advance` refers to `updating`, just step-by-step.

File: tracing/tracer.py [hunk 10]
```diff
@@ -43,12 +63,48 @@ def active_trace(self) -> typing.Iterator[None]:
         finally:
             self.stop_trace()
 
-    def _reset_members(self) -> None:
-        """Resets the variables of the tracer."""
-        self.trace_data = pd.DataFrame(
-            columns=constants.TraceData.SCHEMA.keys()
-        ).astype(constants.TraceData.SCHEMA)
-        self.old_values_by_variable_by_function_name = {}
+    def _update_optimisations(self, fwm: FrameWithMetadata) -> None:
```
[the remark this answers is not in the document]
Comment (bengsparks) [reply tracing/tracer.py:66]: Perhaps `_update_opt_stack` would be better?

File: tests/test_tracer.py [hunk 11]
```diff
@@ -481,11 +490,12 @@ def test_if_tracer_traces_sample_function_with_inner_function_it_collects_correc
     test_object.stop_trace()
 
     actual_trace_data = test_object.trace_data
-    # with pd.option_context("display.max_rows", None, "display.max_columns", None):
-    #    print(actual_trace_data.head(n=20))
-    #    print(actual_trace_data.dtypes)
-    #    print(expected_trace_data.dtypes)
-    assert expected_trace_data.equals(actual_trace_data)
+    with pd.option_context("display.max_rows", None, "display.max_columns", None):
```
[the remark this answers is not in the document]
Comment (bengsparks) [reply tests/test_tracer.py:493]: Done.

**Review by RecurvedBow — APPROVED** [review]

**Pull request merged**